avalon/tools/cbsceneinventory/model.py

```diff
@@ -65,20 +95,99 @@ def data(self, index, role):
 
         return super(InventoryModel, self).data(index, role)
 
-    def refresh(self):
+    def set_hierarchy_view(self, state):
+        """Set whether to display subsets in hierarchy view."""
+        state = bool(state)
+
+        if state != self._hierarchy_view:
+            self._hierarchy_view = state
+
+    def refresh(self, selected=None):
         """Refresh the model"""
 
         host = api.registered_host()
-
-        items = []
-        containers = host.ls()
-        for container in containers:
-            items.append(container)
+        items = host.ls()
 
         self.clear()
-        self.add_items(items)
 
-    def add_items(self, items):
+        if self._hierarchy_view and selected:
+
+            if not hasattr(host.pipeline, "update_hierarchy"):
+                # If host doesn't support hierarchical containers, then
+                # cherry-pick only.
+                self.add_items((item for item in items
+                                if item["objectName"] in selected))
+
+            # Update hierarchy info for all containers
+            items_by_name = {item["objectName"]: item
+                             for item in host.pipeline.update_hierarchy(items)}
+
+            selected_items = set()
+
+            def walk_children(names):
+                """Select containers and extend to chlid containers"""
+                for name in [n for n in names if n not in selected_items]:
+                    selected_items.add(name)
+                    item = items_by_name[name]
+                    yield item
+
+                    for child in walk_children(item["children"]):
+                        yield child
+
+            items = list(walk_children(selected))  # Cherry-picked and extended
+
```

Comment:
Here's the heavy lifting part, since the hierarchy has been cherry-picked, in order to draw the graph, we need to scan all the containers, to see which is the final hierarchy and which is not.

This is the part that we need `"children"` data from host, for filtering.

And the `"parent"` data is required down below, for building the view.